### ISSUE TITLE
Adding Attributes Field to Constructors of Types that are Instances of HasAttributes

### ISSUE THREAD

**Issue by niazim3** [JacquesCarette/Drasil#514]
Related to issues #497 and #408.

@samm82 :

- [ ] AssumpChunk.hs
- [ ] GenDef.hs 
- [ ] ReqChunk.hs

@elwazana :

- [ ] Change.hs
- [ ] Goal.hs

@niazim3 : 

- [ ]  Eq.hs
- [ ]   PhysicalSystemDescription.hs

**Comment by niazim3** [comment]
`psd` in [PhysSystDesc](https://github.com/JacquesCarette/Drasil/blob/cad5e614b0cfa68ce05f69ef7944bdfa53df77a6/code/Language/Drasil/Chunk/PhysSystDesc.hs#L32-L34) has a comment saying that this is a smart constructor without any explicit 'Attributes', so no change was made to this constructor definition.

@szymczdm Both `psd` and `psd'` are not used anywhere at the moment, can they be removed?

**Comment by szymczdm** [comment]
Those should not be removed, they haven't been "plugged in" yet.

**Comment by samm82** [comment]
@szymczdm For adding attributes to constructors that don't have them, how do we identify which constructors do and do not need attributes as an argument?

**Comment by niazim3** [comment]
@szymczdm Are we to add the attributes field to all of the chunks for consistency (mentioned below)?

https://github.com/JacquesCarette/Drasil/blob/e82da482309339c787ee028e38dbb3325569b506/code/Language/Drasil/Chunk/Goal.hs#L22-L25

If so, would you like us to add a FIXME comment wherever we are unsure if Attributes are useful for a particular data type?

**Comment by niazim3** [comment]
@szymczdm @JacquesCarette 
Given the current definition of `tc` in Theory.hs: https://github.com/JacquesCarette/Drasil/blob/e82da482309339c787ee028e38dbb3325569b506/code/Language/Drasil/Chunk/Theory.hs#L69-L72, the `qw` constructor from Quantity.hs https://github.com/JacquesCarette/Drasil/blob/e82da482309339c787ee028e38dbb3325569b506/code/Language/Drasil/Chunk/Quantity.hs#L36-L37 is used in the fourth argument to produce a list of `QuantityDict` by applying it as a function on `q`, which is a `[Quantity]` type.

In order to incorporate Attributes to the `Quantity` chunk, I've updated the definition of `qw` (on my local machine) to 
```
qw :: Quantity q => q -> Attributes -> QuantityDict
qw q atts = QD (nw q) (q^.typ) (symbol q) (getUnit q) atts
```
.

I've been receiving the following error: 
```
C:\Users\Owner\literate-scientific-software\code\Language\Drasil\Chunk\Theory.hs:78:35: error:      
          • Couldn't match expected type ‘[QDefinition] -> [QuantityDict]’      
                        with actual type ‘[Attributes -> QuantityDict]’      
          • The function ‘map’ is applied to three arguments,      
            but its type ‘(q -> Attributes -> QuantityDict)      
                          -> [q] -> [Attributes -> QuantityDict]’      
            has only two      
            In the fourth argument of ‘TC’, namely ‘(map qw q atts)’      
            In the expression: TC cid t s (map qw q atts) (map cw c) atts
```
and I'm unsure how to fix it. I've tried zipping up `atts` and `q` to be passed in to `qw` in the fourth argument and also tried doing `map qw q atts` (with parentheses moved around), but those attempts didn't work. So essentially I guess what I'm asking is how do we map a constructor to a list when a constructor is expecting 2 or more arguments?

**Comment by halonazhao** [comment]
https://github.com/JacquesCarette/Drasil/blob/e82da482309339c787ee028e38dbb3325569b506/code/Language/Drasil/Chunk/Eq.hs#L24
https://github.com/JacquesCarette/Drasil/blob/e82da482309339c787ee028e38dbb3325569b506/code/Language/Drasil/Chunk/Eq.hs#L59-L60
https://github.com/JacquesCarette/Drasil/blob/e82da482309339c787ee028e38dbb3325569b506/code/Language/Drasil/Chunk/Eq.hs#L75-L76
The example above would work because the QDefinition itself includes '_att :: Attributes'.

https://github.com/JacquesCarette/Drasil/blob/e82da482309339c787ee028e38dbb3325569b506/code/Language/Drasil/Chunk/Quantity.hs#L24-L25
https://github.com/JacquesCarette/Drasil/blob/e82da482309339c787ee028e38dbb3325569b506/code/Language/Drasil/Chunk/Quantity.hs#L36-L37
But for this, we can not simply add attributes into it because the QuantityDict doesn't have an attribute field in it.

I guess this is why your change failed.

**Comment by szymczdm** [comment]
@niazim3 if you've got a list of QDefinitions and a list of Attributes (that correspond to each QDefinition), then you can just zip the QDefs and Attributes into a pair, then `map (\(q,att) -> qw q att)` over the list of pairs.

**Comment by niazim3** [comment]
@halonazhao Sorry, I should've mentioned that `QuantityDict` already had been updated to have attributes in it : 
```
data QuantityDict = QD { _id' :: IdeaDict
                       , _typ' :: Space
                       , _symb' :: Stage -> Symbol
                       , _unit' :: Maybe UnitDefn
                       , _attribs :: Attributes 
                       }
makeLenses ''QuantityDict
```

**Comment by JacquesCarette** [comment]
I think we should add attributes to things sparingly. For `Quantity` and above, sure, no problem.  Below? It should be an explicit decision for each of them.

For example `NamedIdea` shouldn't have attributes, as far as I can tell.

**Comment by samm82** [comment]
@JacquesCarette So does `NamedIdea` need attributes? Because it's above `Quantity`.

**Comment by JacquesCarette** [comment]
Sorry, above/below is ambiguous.  I meant those things which are "more complex" than `Quantity` (except for `Theory` which is quite different) can have attributes.  Things that are "simpler", such as `NamedIdea`, `CommonIdea`, etc, quite possibly don't.

**Comment by niazim3** [comment]
Current status of this issue: work originating from `Quantity.hs` had led to errors throughout a lot of the files, which are in the process of getting fixed so that make runs result in empty logs; `UncertainQuantity.hs` "is next on the list" of being updated to work well with changes made in preceding files.

**Comment by niazim3** [comment]
Next steps: as mentioned in the comments of commit https://github.com/JacquesCarette/Drasil/commit/27fdf425b1005553a816019578071fbeaa01f07b ^, the next steps would be to update the constructors in `Unital.hs` to take in attributes as an argument instead of relying on the `HasAttributes` instance (since `UnitalChunk` are built using `Convar`s, which are built using `ConceptChunk`s, which are neither instances of `HasAttributes` nor take in a list of `Attributes` as a field).

**Comment by samm82** [comment]
@JacquesCarette Right now, `ConVar` doesn't have attributes, which came up in commit dbd4b7f1 when @niazim3 and I were updating the GamePhysics Example. Instead of adding attributes to `ConVar`, would it make more sense to build the `ConstrainedChunk` from a `DefinedQuantityDefinition`, since `ConVar` and `DQD` both have a `ConceptChunk`, `Space`, and `Symbol`?

**Comment by szymczdm** [comment]
It appears ConVar and DefinedQuantityDict are identical in terms of the class instances they implement. The only difference is in the way they're constructed. If anything it might be worthwhile to remove one of them entirely since they're not really saying anything different.

**Comment by niazim3** [comment]
It may be better to remove `ConVar` than to remove `DQD`, as it only has `ConceptChunk`, `Symbol`, and `Space`; whereas `DQD` has `Quantity` (which has `unit`, `attribs`, `idea`, `space`, and `symbol`) and `ConceptChunk`. Shall I create an issue for this?

**Comment by szymczdm** [comment]
An issue would be good, and I was thinking the same although we may want to change the way DQD is constructed to avoid the 'double uid' issue mentioned in the comments inside DefinedQuantity.hs.

Also, just to be clear, ConVar and DQD have exactly the same class instances. While ConVar always returns `Nothing` for the `Maybe Unit`, whereas DQD will sometimes have a non-Nothing value, they are databases of the same types of information.

**Comment by samm82** [comment]
@szymczdm Thanks to #537, should this issue now be closed?

**Comment by szymczdm** [comment]
@samm82 Yup, since we've figured out what we're doing with Attributes now (ie. removing them and creating new lenses), I'll go ahead and close this.
